### DISCUSSION TITLE
APA theme and helpers

### DISCUSSION THREAD

**Discussion by davidgohel** [davidgohel/flextable#426] [Ideas]
How to help with APA subject

- A theme function or a something() |> as_flextable() ?
- utilities?
- support for rempsyc

@rempsyc @Buedenbender

guidelines and references:

- https://github.com/rempsyc/rempsyc
- https://apastyle.apa.org/style-grammar-guidelines/tables-figures/tables.

**Comment by Buedenbender** [comment]
Hi in shameless self-promotion, I am quite happy with my [`format_flextable()`](https://github.com/Buedenbender/datscience/blob/main/R/format_flextable.R) function ([doc](https://buedenbender.github.io/datscience/reference/format_flextable.html)).
If this is desireable we could easily use the code of the function for a improved / better version and include it in flextable. Also the [`nice_table()`](https://github.com/rempsyc/rempsyc/blob/main/R/nice_table.R) function of @rempsyc could be a useful basis. 

I think overall a function or template for the APA Style is very useful and would be highly appreciated by a wide community, as there is a whole range of Journals in Social Science and Medicine that explicitly require APA Style. In the APA Publication Manual the Chapter would be "7 Tables and Figures" (pp 195. - 209). There are also examples included, which are also on the apastyle webpage you referenced (examples can be found: [here](https://apastyle.apa.org/style-grammar-guidelines/tables-figures/sample-tables)).
Albeit I can not share the respective chapter from the publication manual publicly (due to (C)), I condensed some important information below. Also I could send you parts of my APA Publication manual via Mail if interested.

PS: Just noticed that I did not credit `flextable` in my function (apologizes, will fix that soon, but included references / credit elsewhere in my package e.g., [here](https://buedenbender.github.io/datscience/articles/flex_table1.html)); [fixed it](https://github.com/Buedenbender/datscience/commit/89dd85799ebc25ca4b0873e018116af6ea182e4b)

**List of Details on Formatting (relevant to flextable):**

- 7.10 Each table should be enumerated in bold face (e.g. **Table 1**), followed by
- 7.11 Table heading (in sentence case and in italic
- 7.12 Column Headings, should be centered above their column, use singular 
- 7.13 Body: Line spacing can be variable either (single spaced, one-and-a-half spaced, double spaced). Entries longer than one line (0.15 in indent), cell entries flush left alignment, (if entries significantly shorter than their heading its acceptable to center)
- 7.14 Table Note: There are three type of notes _general notes_, which always start with the word _Note._ in cursive. general notes qualify, explain and provide further information on the table (e.g. explains abbreviations, or copyright attributions); _Specific Notes_ refer to a particular cell, column, or row and are usualy denoted by superscript lowercase letter; _Probability Note_ describes how asterisks are used to indicate significance for _p_ values. Each type of note should start on a new line

**Comment by rempsyc** [comment]
Ok. Here's a first attempt.

``` r
library(flextable)
theme_apa <- function(x) {
  if (!inherits(x, "flextable")) {
    stop("theme_apa supports only flextable objects.")
  }
  apa.border <- list("width" = 1, color = "black", style = "solid")
  x <- font(x, part = "all", fontname = "Times New Roman")
  x <- line_spacing(x, space = 2, part = "all")
  x <- hline_top(x, part = "head", border = apa.border)
  x <- hline_bottom(x, part = "head", border = apa.border)
  x <- hline_top(x, part = "body", border = apa.border)
  x <- hline_bottom(x, part = "body", border = apa.border)
  x <- align(x, align = "center", part = "all")
  x <- valign(x, valign = "center", part = "all")
  x <- colformat_double(x, digits = 2)
  fix_border_issues(x)
}
head(mtcars*22.22) |>
  flextable() |>
  theme_apa()
```
![](https://i.imgur.com/Nwy6wkC.png)
<sup>Created on 2022-07-21 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

## Discussion

1. Initially, I wrote the function with the native pipe operator `|>` but realized that you don’t have a `Depends` section specifying the minimal R version required in your `DESCRIPTION` file. So I assumed `flextable` doesn't support the native pipe yet and rewrote the code accordingly.
2. Few things I’ve realized: size 12 and `theme_booktabs` are default now (has this changed or always was? Had the feeling `theme_vanilla` was the default before) so not needed so I’ve removed them. Not sure if it’s safer to specify explicitly still, in case defaults change in the future. Following David’s comment, I removed `height=0.55` and instead used double space along `valign`, which allowed me to remove `hrule`.
3. Not sure about including `set_table_properties(layout = "autofit")` or not, but I think default is fine since it’s too situational.
6. The caption (“Table 1” on first line, “Description of table” on second line), note, footnotes, etc. would be required for APA style. However, to stay consistent with other templates, I am not sure it should be integrated with the APA template, or whether it should be required in a second step. Technically, we are, after all, writing a theme, not a full-fledged function. So I would tend to go for the basic APA template no caption, and then just _maybe_ add a function to add the double caption (and notes/footnotes) as per @Buedenbender's way, if at all. I would be quite happy with that to start with. Up for discussion of course.
4. The default digit is tricky because it depends on the statistic (p value is 3 digits, r values are 2 digits, descriptive stats are 1 digit) and this is what `rempsyc::nice_table` deals with, but I think a safe default would be 2 digits. If people really want specific digits, they can manually specify after.
5. David if you’d be interested, perhaps there would be a possibility of some overlap between `nice_table` and `theme_apa` for basic statistical formatting (perhaps italics for common statistical symbols, combination of upper and lower bounds of the 95% confidence interval, etc.), however the function is pretty comprehensive, so I don’t know how far you’d like to go into that.

@Buedenbender, anything missing in your opinion? I tried combining both our custom functions. Thanks for setting up all the resources and explanations in your first post!

**Reply by Buedenbender** [reply]
Very well done. I agree footer and header might not need to be part of the basic theme as this is not a full function.
One possible addition I'd like to suggest could be instead of stoping when not a flextable is provided, the function could try to convert and if this is not possible throw the same message

e.g.,

```r
# Conversion if possible (e.g., for data.frame)
  if (!inherits(x, "flextable")) {
    tryCatch(
      x <- flextable::flextable(x),
      error = function(c) stop("Error could not convert to a flextable object. Stopping - theme_apa() only supports flextable objects.")
    )
  }
```

I am not quite sure if this is proper behavior for a theme, as it might, should only take flextable objects. Still I believe it would not hurt, to include this one line. E.g., for early data exploration this would enable me to call `theme_apa(iris)` and it would work fine.

**Comment by rempsyc** [comment]
@davidgohel would it be useful if I submitted a PR with the new APA theme?

**Comment by davidgohel** [comment]
(Sorry for the delay; I was in holidays, and then I tested flextable with quarto and since I am working on captions... but it is related to our subjects)

You can prepare a PR, it will be welcome. But we may have to iterate over it, because there is something to enable for captions (with the caption structure table xx/caption/table), it must be made compatible with cross-referencing in bookdown and in quarto if possible. I am not sure if it's easy. (Everything related to caption is crazily time-consuming!)

I am also working on enabling captions formatting as you have suggested. That's almost ready.

**Comment by davidgohel** [comment]
Sorry again for the delay but I do what I can :)


@rempsyc to answer your questions

1. we don't use pipes because we want user that only have old versions can use the package latest versions. Not sure if it often happens but that's a decision asked by a customer a long time ago so we stick to it :)
2. a problem only occurs with pdf that does not support line height nor height. I am trying to find a solution, I doubt it is possible easily. 
3. Agree
4. Agree. Below a suggestion to start...
5. I prefer to not redo your work on these points 
6. I may not have enough time. If we can provide the tools you need, it will be a very first good step :) 

--------

Starting from the PR of @rempsyc and the latest GitHub version of flextable,
here is a function to add the captions as defined in APA rules. Feel free to comment or suggest improvements. 


```
library(officer)
library(flextable)

apa_paragraph_tabnum <- function (z) {
  
  pr <- fp_text_default(font.size = 12, font.family = "Times New Roman", bold = TRUE)
  
  run_str_pre <- as_chunk(z$pre_label, props = pr)
  
  seq_str <- paste0("SEQ ", z$seq_id, " \u005C* Arabic")
  if(!is.null(z$start_at) && is.numeric(z$start_at)){
    seq_str <- paste(seq_str, "\\r", as.integer(z$start_at))
  }
  run_word_field <- as_word_field(x = seq_str, prop = pr)
  
  run_word_styleref_1 <- NULL
  run_word_styleref_2 <- NULL
  if(z$tnd > 0){
    run_word_styleref_1 <- as_word_field(x = paste0("STYLEREF ", z$tnd, " \\r"), prop = pr)
    run_word_styleref_2 <- as_chunk(z$tns, prop = pr)
  }
  
  args <- list(run_str_pre, run_word_styleref_1, run_word_styleref_2, run_word_field)
  args <- Filter(function(z) !is.null(z), args)
  do.call(as_paragraph, args)
}

apa_decorate <- function(x, tab_caption, table_num = NULL) {
  
  table_num_fp <- fp_text_default(font.size = 12, font.family = "Times New Roman", bold = TRUE)
  
  x <- add_header_lines(x, values = c("", ""))
  x <- align(x, i = 1:2, part = "header", align = "left")
  if (is.null(table_num)) {
    z <- flextable:::get_word_autonum(x, tab_props = opts_current_table())
    x <- compose(x, i = 1, j = 1, part = "header", value = apa_paragraph_tabnum(z))
  } else {
    x <- compose(x, i = 1, j = 1, part = "header", value = as_paragraph(as_chunk(table_num, props = table_num_fp)))
  }
  
  x <- compose(
    x = x, i = 2, 
    j = 1, part = "header", 
    value = as_paragraph(
      as_chunk(tab_caption, props = fp_text_default(font.size = 12, font.family = "Times New Roman", italic = TRUE))
    )
  )
  x <- line_spacing(x, i = 2, part = "header", space = 2)
  x <- valign(x, i = 2, part = "header", valign = "center")
  
  x <- border(x = x, i = 1:2, border = fp_border_default(width = 0), part = "header")
  x <- hline(x = x, i = 2, border = fp_border_default(width = 1, color = "black"), part = "header")
  
  x <- font(x, fontname = "Times New Roman", part = "header")
  x <- fontsize(x, size = 12, part = "header")
  
  x
}

theme_apa <- function(x, line_spacing = 2, ...) {
  if (!inherits(x, "flextable")) {
    stop("theme_apa supports only flextable objects.")
  }
  apa.border <- list("width" = 1, color = "black", style = "solid")
  x <- font(x, part = "all", fontname = "Times New Roman")
  x <- fontsize(x, size = 12, part = "all")
  x <- line_spacing(x, space = line_spacing, part = "all")
  x <- hline_top(x, part = "head", border = apa.border)
  x <- hline_bottom(x, part = "head", border = apa.border)
  x <- hline_top(x, part = "body", border = apa.border)
  x <- hline_bottom(x, part = "body", border = apa.border)
  x <- align(x, align = "center", part = "all")
  x <- valign(x, valign = "center", part = "all")
  x <- colformat_double(x, digits = 2)
  fix_border_issues(x)
}

ft_1 <- head(cars) |>
  flextable() |>
  set_caption(autonum = run_autonum(seq_id = "tab", bkm = "mtcars", prop = fp_text_default(bold = TRUE)))|>
  theme_apa()|>
  # apa_decorate(tab_caption = "qsdq qdqsd qdqsd")|>
  apa_decorate(table_num = "Table 1", tab_caption = "qsdq qdqsd qdqsd")|>
  fix_border_issues()

print(ft_1, preview = "docx")
print(ft_1, preview = "html")
print(ft_1, preview = "pptx")
print(ft_1, preview = "pdf", latex_engine = "xelatex")

```


This solution comes with some limits:

1. not cross-reference-able with bookdown and quarto!
2. PDF line height (2 for the caption) is not possible for now
3. Table numbers will be made by hand, that's not ideal. Maybe a pandoc filter could be used...

**Reply by rempsyc** [reply]
Awesome! Thanks for working on this.

1. I'm getting an error in this reprex
```
Error in set_caption(flextable(head(cars)), autonum = run_autonum(seq_id = "tab",  : 
  argument "caption" is missing, with no default
```
Then I updated to current github version, and got this instead:
```
Error in run_autonum(seq_id = "tab", bkm = "mtcars", prop = fp_text_default(bold = TRUE)) : 
  could not find function "run_autonum"
```
It works without the `set_caption` line though.

2. I noticed you changed the default line spacing from 2 to 1.2, but it would be preferable to keep 2 as the default, as this is what people using APA style would expect.

3. The table caption (decoration) is not in Times New Roman size 12 anymore, but Arial 11.

4. In `rempsyc::nice_table`, I also have this line, `fontsize(part = "all", size = 12)`, which I thought wasn't necessary for `theme_apa` because last I tested, 12 was now the default font size. But now it seems it's back to 11? So we would need to add back that line to `theme_apa`. It would be better to specify explicitly in case defaults change again in the future.

5. The top line is gray instead of black like the others.

**Reply by rempsyc** [reply]
Also I'm not sure I understand your last point:

> Table numbers will be made by hand, that's not ideal. Maybe a pandoc filter could be used...

Does that mean that the table can't be formatted based on numeric values anymore after the theme has been applied?

**Reply by davidgohel** [reply]
1. I updated the code by adding `library(officer)`
2. code updated
3. code updated
4. code updated (I think default size has not been different from 11 since always, at least since 2 years: https://githistory.xyz/davidgohel/flextable/blob/master/R/defaults.R.)
5. code updated

7. That mean you have to write "Table 2" for HTML, PDF and PPTX. For DOCX you can benefit from `run_autonum()` defined with `set_caption()` (uncomment the commented `apa_decorate ` line and comment the uncommented `apa_decorate` line to test). 

Where do you want to use it? R Markdown, Bookdown, Quarto, only officer? Unfortunately, the first is not supposed to provide numbered captions; the second must create the "table 1" values (so some features of flextable must be forgotten), the third has not yet support for table captions in Word but as bookdown, chunk formats are lost. 

Caption is a nightmare. Cross-references and auto numbering are major issues, rmarkdown/bookdown/quarto do not expect captions in tables as recommended in apastyle.apa.org. I have implemented as you suggested options for paragraph settings, also captions can now be made of blue chunks, bold chunks and italic chunks but Quarto do not keep them as is. Here, everything will be inside the table, but we lose the ability to get scanned by quarto or bookdown so that all tables captions are well numbered.

**Reply by rempsyc** [reply]
1. In HTML, the cell values appear centered, as they should. But in Word, the values appear aligned toward the top. It seems like this is not unique to `theme_apa`, because even `rempsyc::nice_table` is affected by this. Edit: actually I tried downloading older versions of `flextable` and the same issue was already there (probably always was). I remember having to deal with this before (why I was using `height=0.55` etc. instead of `valign`). Maybe I just thought it was fixed by looking at the HTML and forgot to check the Word documents. Anyway, that's not a big deal, I guess we can't have both double space and perfect cell centering because of limitations of Word.

2. FYI, the top line does not appear in Word anymore in `rempsyc::nice_table` , even though it uses the hidden `add_header_lines` and not actual captions. Reprex:

``` r
library(rempsyc)
my_table <- nice_table(mtcars[1:3, ],
                       title = "Motor Trend Car Road Tests",
                       footnote = "Note: 1974 Motor Trend US magazine.")
# html
my_table
```
![](https://i.imgur.com/0ASzNeH.png)
``` r
print(my_table, preview = "docx")
```
![](https://i.imgur.com/Nv8A3Sa.png)
``` r
packageVersion("flextable")
#> [1] '0.7.4.8'
```
<sup>Created on 2022-09-02 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

With older version, in Word we get
``` r
packageVersion("flextable")
#> [1] ‘0.7.1’
```
![](https://i.imgur.com/kj5cUzP.png)

If that is an expected change, that means I'll have to adapt my code and force usage of `flextable (>= 0.7.4.8)` (probably only once the new version makes it to CRAN to make it compatible for most users). [Edit: working on a self-contained minimal example right now]

> That mean you have to write "Table 2" for HTML, PDF and PPTX. For DOCX you can benefit from run_autonum() defined with set_caption() (uncomment the commented apa_decorate line and comment the uncommented apa_decorate line to test).

That's no big deal I think, most people won't care that much about this. Plus most people will use this for Word.

> Where do you want to use it? R Markdown, Bookdown, Quarto, only officer?

APA style is used for scientific publishing, and people using `LaTeX` etc. are still a minority. That means 99% of people will use Word, and 1% might try to save to PPT or PDF, but just for fun. I however will use HTML, but only for package vignettes (relying on `rmarkdown`) showing how to save to Word, on my `pkgdown` site. If not for scientific publishing (in APA journals), people can use the format they want so it's less of a problem.

**Reply by davidgohel** [reply]
1. yeah, I can see Word and PowerPoint don't really handle that correctly (I did try manually) unless there is a certain height. I would be happy to change that if possible, but I am not sure if it is possible...
2. This is discussed in the opened issue #450, if your package uses `hline()` instead of `border()`, it should be ok
3. OK, as you know there is an ongoing work on captions, last missing functionality if for quarto/docx
4. OK, I guess most people will use 'quarto', 'rmarkdown' or 'bookdown'

**Reply by rempsyc** [reply]
I tried your updated example again and it looks really neat now! :) I wonder if that would be a go for the first version. We could always make minor tweaks when captions get fully supported in a future `flextable` version.

**Reply by davidgohel** [reply]
almost :) With the new caption features (again you need to update 'flextable' package from github), we may have no need for `apa_decorate()`! I would be far more happy with this solution. 

See the following quarto example: 


`````
---
title: "APA theme"
format: 
  docx:
    keep-md: true
---

```{r include=FALSE}
library(flextable)
library(officer)
knitr::opts_chunk$set(echo = FALSE)
fpp <- fp_par(text.align = "left", padding = 3, line_spacing = 2)
apa_fp_text <- fp_text(font.size = 12, font.family = "Times New Roman")

theme_apa <- function(x, line_spacing = 2, ...) {
  if (!inherits(x, "flextable")) {
    stop("theme_apa supports only flextable objects.")
  }
  apa.border <- list("width" = 1, color = "black", style = "solid")
  x <- font(x, part = "all", fontname = "Times New Roman")
  x <- fontsize(x, size = 12, part = "all")
  x <- line_spacing(x, space = line_spacing, part = "all")
  x <- hline_top(x, part = "head", border = apa.border)
  x <- hline_bottom(x, part = "head", border = apa.border)
  x <- hline_top(x, part = "body", border = apa.border)
  x <- hline_bottom(x, part = "body", border = apa.border)
  x <- align(x, align = "center", part = "all")
  x <- valign(x, valign = "center", part = "all")
  x <- colformat_double(x, digits = 2)
  fix_border_issues(x)
}

```

blah blah blah `r run_reference(id = "tbl-flextable-1")`.

```{r}
#| ft.align: left
flextable(mtcars[1:4, 1:3]) |>
  theme_apa() |> 
  set_caption(
    caption = as_paragraph(
      as_chunk("\nCaption", props = apa_fp_text)
    ),
    autonum = run_autonum(seq_id = "tab", bkm = "tbl-flextable-1", prop = apa_fp_text),
    align_with_table = FALSE,
    word_stylename = "Table Caption",
    fp_p = fpp
  )
```

`````


It produces 
[Untitled.docx](https://github.com/davidgohel/flextable/files/9501218/Untitled.docx)

<img width="889" alt="Capture d’écran 2022-09-07 à 00 21 18" src="https://user-images.githubusercontent.com/4331618/188750661-689f4285-2561-4402-ba07-3a8b8c12bfdf.png">


If you like this solution, we can merge your PR without any other modification.

**Reply by rempsyc** [reply]
Just downloaded `quarto` and tested both your examples. They work well, congrats and thanks for your work on this! I was wondering how you set the style for the caption because at first I had somehow missed this line with the new `fp_text` function. That's cool.

Also, not that it bothers me, though I don't know if it's normal, but when I click on the caption and check styles, I don't see the caption "Table Caption" style:

![](https://i.imgur.com/pPnbYlS.png)

Another thing. If using the default, I am not able to make the first level header bold and remove the colon. Is this all automatized? Seems like the default style overwrites the header style here.

````
---
title: "APA theme"
format: 
  docx:
    keep-md: true
---

```{r include=FALSE}
library(flextable)
library(officer)
knitr::opts_chunk$set(echo = FALSE)
fpp <- fp_par(text.align = "left", padding = 3, line_spacing = 2)

theme_apa <- function(x, line_spacing = 2, ...) {
  if (!inherits(x, "flextable")) {
    stop("theme_apa supports only flextable objects.")
  }
  apa.border <- list("width" = 1, color = "black", style = "solid")
  x <- font(x, part = "all", fontname = "Times New Roman")
  x <- fontsize(x, size = 12, part = "all")
  x <- line_spacing(x, space = line_spacing, part = "all")
  x <- hline_top(x, part = "head", border = apa.border)
  x <- hline_bottom(x, part = "head", border = apa.border)
  x <- hline_top(x, part = "body", border = apa.border)
  x <- hline_bottom(x, part = "body", border = apa.border)
  x <- align(x, align = "center", part = "all")
  x <- valign(x, valign = "center", part = "all")
  x <- colformat_double(x, digits = 2)
  fix_border_issues(x)
}

```

```{r}
#| ft.align: left
apa_fp_text <- fp_text(font.size = 12, font.family = "Times New Roman")
flextable(mtcars[1:4, 1:3]) |>
  theme_apa() |> 
  set_caption(
    caption = as_paragraph(
      as_i(as_chunk("\nItalic Caption", props = apa_fp_text))
    ),
    autonum = run_autonum(seq_id = "tab", bkm = "tbl-flextable-1", prop = apa_fp_text),
    align_with_table = FALSE,
    word_stylename = "Table Caption",
    fp_p = fpp
  )
```

---

```{r}
#| ft.align: left
apa_fp_text <- fp_text(font.size = 12, font.family = "Times New Roman")
apa_fp_text2 <- fp_text(font.size = 12, font.family = "Times New Roman",
                        bold = TRUE)

flextable(mtcars[1:4, 1:3]) |>
  theme_apa() |> 
  set_caption(
    caption = as_paragraph(
      as_b("Table 1"),
      as_i(as_chunk("\nItalic Caption", props = apa_fp_text))
    ),
    autonum = run_autonum(seq_id = "tab", bkm = "tbl-flextable-1", prop = apa_fp_text),
    align_with_table = FALSE,
    word_stylename = "Table Caption",
    fp_p = fpp
  )
```
````
![](https://i.imgur.com/IIiyDli.png)

**Reply by davidgohel** [reply]
> no 'table caption'

I can't reproduce, did you update the package with latest version?

<img width="1020" alt="Capture d’écran 2022-09-07 à 09 38 52" src="https://user-images.githubusercontent.com/4331618/188827922-cb358c5a-a197-4b0f-bcdf-ad7b6027e876.png">

> remove the colon

thanks, fixed, https://github.com/davidgohel/flextable/commit/88d83f8e07299140eb6ee15575506765ffba989e

Use `post_label = ""`: 

```
autonum = run_autonum(seq_id = "tab", bkm = "tbl-flextable-1", prop = apa_fp_text, post_label = "")
```

You could also have used chunk option `tab.cap.sep`. I now prioritize values from caption over values from chunk options (section 'Table caption' in `?knit_print.flextable`)


-----
there:

```
caption = as_paragraph(
      as_b("Table 1"),
      as_i(as_chunk("\nItalic Caption", props = apa_fp_text))
    ),
    autonum = run_autonum(seq_id = "tab", bkm = "tbl-flextable-1", prop = apa_fp_text),
```

`autonum` is generating the "table xxx" but you added it again with `as_b("Table 1")`. Quarto is supposed to add number to captions. That's how it works. If you leave quarto, you can switch to a manual version.

```
caption = as_paragraph(
      as_b("Table 1"),
      as_i(as_chunk("\nItalic Caption", props = apa_fp_text))
    ),
    autonum = NULL,
```

Again (not ranting but explaining the complexity...) captions are not managed the same way in bookdown in quarto and in rmarkdown.

Have some tests with only kable, maybe it will help.

**Reply by rempsyc** [reply]
> Use post_label = "":

Cool! Works.

> autonum is generating the "table xxx" but you added it again with as_b("Table 1").

1. Ok I first tried removing the `autonum` argument, or setting it to `null` explicitly, which removes the number, but it doesn't remove the "Table" part. How can I control that to remove if needed?

![](https://i.imgur.com/iC3nJFo.png)

2. If I want to use automatic numbering with `autonum`, but want that line to be bold, how can I do it? (for APA style)

> I can't reproduce, did you update the package with latest version?

3. Latest versions of `officer` and `flextable`, but I'm using Office 365

![word version](https://user-images.githubusercontent.com/13123390/188939987-5197d2ba-6b70-4084-932a-d0823df779e7.png)

4. There is a `?(caption)` bit appearing below the tables for some reason.

![caption](https://user-images.githubusercontent.com/13123390/188935374-3bf791bb-ad22-41c5-a4c2-620d47cb55bb.png)

5. I always get this message when I open the Word documents:

![other fields](https://user-images.githubusercontent.com/13123390/188935454-6376f7b3-71cb-419c-953f-f449f1fe9cea.png)

6. The script below produces two tables for me currently.

````
---
title: "APA theme"
format: 
  docx:
    keep-md: true
---

```{r include=FALSE}
library(flextable)
library(officer)
knitr::opts_chunk$set(echo = FALSE)
fpp <- fp_par(text.align = "left", padding = 3, line_spacing = 2)

theme_apa <- function(x, line_spacing = 2, ...) {
  if (!inherits(x, "flextable")) {
    stop("theme_apa supports only flextable objects.")
  }
  apa.border <- list("width" = 1, color = "black", style = "solid")
  x <- font(x, part = "all", fontname = "Times New Roman")
  x <- fontsize(x, size = 12, part = "all")
  x <- line_spacing(x, space = line_spacing, part = "all")
  x <- hline_top(x, part = "head", border = apa.border)
  x <- hline_bottom(x, part = "head", border = apa.border)
  x <- hline_top(x, part = "body", border = apa.border)
  x <- hline_bottom(x, part = "body", border = apa.border)
  x <- align(x, align = "center", part = "all")
  x <- valign(x, valign = "center", part = "all")
  x <- colformat_double(x, digits = 2)
  fix_border_issues(x) 
}

```

```{r}
#| echo: true
packageVersion("officer")
packageVersion("flextable")
```

This should not be italic `r run_reference(id = "tbl-flextable-1")`. \n \n

---

```{r}
#| label: tbl-flextable-1
#| ft.align: left
apa_fp_text <- fp_text(font.size = 12, font.family = "Times New Roman")
flextable(mtcars[1:4, 1:3]) |>
  theme_apa() |> 
  set_caption(
    caption = as_paragraph(
      as_i(as_chunk("\nItalic Caption", props = apa_fp_text))
    ),
    autonum = run_autonum(seq_id = "tab", bkm = "tbl-flextable-1", 
                               prop = apa_fp_text, post_label = ""),
    align_with_table = FALSE,
    word_stylename = "Table Caption",
    fp_p = fpp
  )
```
````

**Reply by davidgohel** [reply]
1. You added a label. My example did not contain label. The rules to follow in quarto are not clear to me. Because there is a label, quarto adds it after the printed object. This is quarto and I can't do anything about it. let's wait Quarto manage the captions in Word. If you add a label and `tbl-cap`, things are as expected (by me)
2. See `?run_autonum` and have a look at `prop`.
3. thanks, AFAIK Word version does not matter. The docx format is stable. 
Could you run this and post the results: `officer::docx_summary(read_docx("your docx file"))`. Of course change "your docx file" to the correct path to your docx file.
4. same reason than 1. Quarto is seeing a label but no `tabl-cap`. 
5. see https://ardata-fr.github.io/officeverse/faq.html#update-fields
6. Because you added a label

**Reply by davidgohel** [reply]
I will update doc now. I did what was possible with quarto/bookdown/rmarkdown/officer. I think that with officer it is ok. With bookdown, you can get frustrations (not possible to format the caption paragraph) but overall is really nice and consistant, with quarto that is still *in dev*, we should wait

**Reply by rempsyc** [reply]
1. Thanks! Yeah I think I tried different things and tried adding back the label from the other example from the other issue. Works now after removing the label :)
2. Arg, that was it. I was changing the bold in `apa_fp_text` so it was changing it at both places! I had to create two different ones. Works now! :D

3. Below:

``` r
library(officer)
docx_summary(read_docx("flextable test2.docx"))
#>      doc_index content_type    style_name
#> 1            1    paragraph         Title
#> 2            2    paragraph   Source Code
#> 3            3    paragraph   Source Code
#> 4            4    paragraph   Source Code
#> 5            5    paragraph   Source Code
#> 6            6    paragraph Table Caption
#> 1.1          7   table cell          <NA>
#> 1.4          7   table cell          <NA>
#> 1.7          7   table cell          <NA>
#> 1.10         7   table cell          <NA>
#> 1.13         7   table cell          <NA>
#> 2.2          7   table cell          <NA>
#> 2.5          7   table cell          <NA>
#> 2.8          7   table cell          <NA>
#> 2.11         7   table cell          <NA>
#> 2.14         7   table cell          <NA>
#> 3.3          7   table cell          <NA>
#> 3.6          7   table cell          <NA>
#> 3.9          7   table cell          <NA>
#> 3.12         7   table cell          <NA>
#> 3.15         7   table cell          <NA>
#>                                        text level num_id row_id is_header
#> 1                                 APA theme    NA     NA     NA        NA
#> 2                 packageVersion("officer")    NA     NA     NA        NA
#> 3                             [1] '0.4.4.6'    NA     NA     NA        NA
#> 4               packageVersion("flextable")    NA     NA     NA        NA
#> 5                            [1] '0.7.4.17'    NA     NA     NA        NA
#> 6    Table SEQ tab \\* ArabicItalic Caption    NA     NA     NA        NA
#> 1.1                                     mpg    NA     NA      1      TRUE
#> 1.4                                   21.00    NA     NA      2     FALSE
#> 1.7                                   21.00    NA     NA      3     FALSE
#> 1.10                                  22.80    NA     NA      4     FALSE
#> 1.13                                  21.40    NA     NA      5     FALSE
#> 2.2                                     cyl    NA     NA      1      TRUE
#> 2.5                                    6.00    NA     NA      2     FALSE
#> 2.8                                    6.00    NA     NA      3     FALSE
#> 2.11                                   4.00    NA     NA      4     FALSE
#> 2.14                                   6.00    NA     NA      5     FALSE
#> 3.3                                    disp    NA     NA      1      TRUE
#> 3.6                                  160.00    NA     NA      2     FALSE
#> 3.9                                  160.00    NA     NA      3     FALSE
#> 3.12                                 108.00    NA     NA      4     FALSE
#> 3.15                                 258.00    NA     NA      5     FALSE
#>      cell_id col_span row_span
#> 1         NA       NA       NA
#> 2         NA       NA       NA
#> 3         NA       NA       NA
#> 4         NA       NA       NA
#> 5         NA       NA       NA
#> 6         NA       NA       NA
#> 1.1        1        1        1
#> 1.4        1        1        1
#> 1.7        1        1        1
#> 1.10       1        1        1
#> 1.13       1        1        1
#> 2.2        2        1        1
#> 2.5        2        1        1
#> 2.8        2        1        1
#> 2.11       2        1        1
#> 2.14       2        1        1
#> 3.3        3        1        1
#> 3.6        3        1        1
#> 3.9        3        1        1
#> 3.12       3        1        1
#> 3.15       3        1        1
```
<sup>Created on 2022-09-07 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

I do have a field on the table number, and when I right-click on it, I can update field, etc.

![](https://i.imgur.com/U9a3Tja.png)

I also have the EndNote integration, so I don't know if that can mess up code fields...

4. Fixed by removing the label also! :)

5. This was actually also caused by the label!

6. Same.

**Reply by rempsyc** [reply]
Only issue I see is that using the automatic label, it appears in Word but not in HTML, which will be problematic in my vignettes if people cannot see the result (unless saving as images and then loading local image).

**Reply by davidgohel** [reply]
See the line 6, it's written "Table Caption" in column `style_name`

For you last/new issue, I still don't have an answer. I suggest to not use HTML and instead use `doconv::to_miniature()` to prepare images of the Word documents.

**Reply by rempsyc** [reply]
Thanks and above all congratulations on all this work and accomplishment! 😄

**Reply by davidgohel** [reply]
Well now, there is much more work to document that 😭 and submit to CRAN. no thanks to you 😉
